### PR TITLE
Add context as input to deepl translation requests

### PR DESCRIPTION
**Pull request by genox**
This is a simple implementation of a context buffer to (hopefully) receive more accurate translations.

### CODE REVIEW

File: src/translate/index.ts [hunk 1]
```diff
@@ -127,9 +128,11 @@ export default async function translate(this: Command) {
       continue;
     }
 
+    let contextBuffer = "";
+
     progress.start(filteredTranslations.length, 0);
     for (let [key, sourceText] of filteredTranslations) {
-      if (options.compiledI18n && Boolean(sourceText) === false) {
+      if (options.compiledI18n && !Boolean(sourceText)) {
```
Comment (tzdesign) [inline src/translate/index.ts:135]: @genox please revert this. The upper one is far more readable.
You can easily oversee the exclamation mark.

File: src/translate/index.ts [hunk 2]
```diff
@@ -148,6 +151,16 @@ export default async function translate(this: Command) {
...
+      const contextToAdd = typeof sourceText === "string" ? sourceText : Object.values(sourceText).join('\n');
+
+      if(contextBuffer.length > maxContextLength) {
+        let newBuffer = contextBuffer.substring(contextToAdd.length);
+        newBuffer += contextToAdd;
+        contextBuffer = newBuffer;
+      } else {
+        contextBuffer += contextToAdd;
+      }
+
```
Comment (tzdesign) [inline src/translate/index.ts:163]: @genox do I understand it correct, that the context will just receive the source value and that's enough?

I actually thought, that the context is more like a standard prompt to tell deepl about the translation.
Do you have a link to the documentation I looked it up, but wasn't able to find it.

**Review by tzdesign — CHANGES_REQUESTED** [review]
@genox thank you very much, for adding it. Would you check the comments?

File: src/translate/index.ts (hunk 2 above)
Comment (genox) [reply src/translate/index.ts:163]: https://www.deepl.com/de/docs-api/translate-text/translate-text

Look for the context request parameter.

It's not a prompt. What you do when you provide context is that the requested string is translated in a context - like each new sentence would be in a larger document. Deepl splits longer texts into sentences and translates them one by one. The entirety of the text is each operations context.

They do write it isn't production ready, but well.. what to do - every other solution is much more complicated.

Comment (tzdesign) [reply src/translate/index.ts:163]: Ah I see. This looks good I will test it on the weekend and merge/deploy it.

I will play around with it. We have already 500+ translations in our qwik app.

**Comment by tzdesign** [comment]
@genox did you had better results?

I experienced the exact output.
What do you think about finding the document where the translation appears as the context?

I tested it with a lot components and single words were finally correct.
Do you think we should make it optional as deepl states it as alpha and has a disclaimer on it?

**Comment by genox** [comment]
@tzdesign Regarding putting it behind an optional flag, sure, that is a sensitive approach.

I think, generally speaking, that the more context can be provided, the more chance there is to get a good translation.

Maybe my approach is flawed, maybe using whitespaces instead of newlines would give a different result. I have no clue about the internals of how this is processed other than the stuff they write in their api docs. I was also considering writing an A4 page worth of text or have chatgpt generate one that outlines the scenario and setting of the project and "sets the tone", so to speak, and use that as a context instead of buffering strings. Or a mix of all approaches.

Did you send the source code as context? What components do you mean?

**Comment by tzdesign** [comment]
Yes mostly the translation were quite equal.

I send the source code within the contexts for a few examples which were wrong and they were significant better. I'm so full of projects, but my plan is to refactor this to send the files concatenated in which the translations occur.

We are actually in a process of auto translating product data on a big scale and leveraging the context. I can tell you more in a few weeks, because I have some native speakers who might verify a large set of translations for me.

**Comment by tzdesign** [comment]
I'm almost done with a grep script, scanning for the current translation location and have very good results. When I am done, I will make it optional to use your solution or the translation +-20 lines as context.

**Comment by tzdesign** [comment]
@genox i added it with grep. The code becomes the context. We have very good experience with this in our project. But the grep part becomes slower in relation to the project size.

**Comment by genox** [comment]
@tzdesign Thank you!

**Pull request closed without merging**